File: 3er CURSO/2-CUATRI/SD/PRACTICAS/Práctica 2_MOD/Cliente_Sudoku/src/cliente_sudoku/Cliente_Sudoku.java
package cliente_sudoku;

import java.rmi.NotBoundException;
import java.rmi.RemoteException;
import java.rmi.registry.LocateRegistry;
import java.rmi.registry.Registry;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;
import libreria_sudoku.GestorJuegosIntf;
import libreria_sudoku.TipoDificultad;

/**
 *
 * @author Ismael Da Palma Fernandez
 */
public class Cliente_Sudoku {

    static Scanner sc_int = new Scanner(System.in);
    static Scanner sc_str = new Scanner(System.in);
    static ArrayList<ListaSudoku> Sudokus = new ArrayList(); //Inicialmente vacio
    static int aviso = 0;
    static String Ayuda = "";
    static int fila_aux, col_aux;

    public static int MenuPrincipal(){
        int valor;
        
        System.out.println("---- Servidor de Juegos ----");
        System.out.println("1.- Gestionar Juegos."); //Llama a MenuGestion
        System.out.println("2.- Jugar."); //Llama a MenuJuego (Si el sudoku esta vacio no deja jugar)
        System.out.println("0.- Salir.");
        System.out.print("\nElige Opción: ");
        valor = sc_int.nextInt();
        
        return valor;
    }
    
    public static int MenuGestion() {
        int valor;
        
        System.out.println("---- Gestión de Juegos ----");
        System.out.println("1.- Crear Juego."); //Llama a MenuDificultad
        System.out.println("2.- Borrar Juego.");
        System.out.println("3.- Seleccionar Juego."); //Llama a ListarSudokus
        System.out.println("0.- Salir."); 
        System.out.print("\nElige Opción: ");
        valor = sc_int.nextInt();

        return valor;
    }
    
    public static int MenuJuego() {
        int valor;
        
        System.out.println("---- Menú de Juegos ----");
        System.out.println("1.- Poner Valor.");
        System.out.println("2.- Borrar Valor.");
        System.out.println("3.- Ayuda.");
        System.out.println("0.- Salir."); 
        System.out.print("\nElige Opción: ");
        valor = sc_int.nextInt();

        return valor;
    }

    public static int ListarJuegos(boolean Borrar){
        int valor;
        System.out.println("\n\nLista de Juegos Almacenados\n");
        System.out.println("Posición        Código          Dificultad");
        System.out.println("------------------------------------------");
        int j = 1;
        for (int i = 0; i < Sudokus.size(); i++) {
            System.out.println(j++ + "\t\t" + Sudokus.get(i).Codigo + "\t\t" + Sudokus.get(i).Dificultad);
        }
        if(Borrar)
            System.out.print("\nIntroduce la posición del Juego a borrar: ");
        else
            System.out.print("\nIntroduce la posición del Juego a continuar: ");
        
        valor = sc_int.nextInt();
        return valor-1;
    }
    
    public static int MenuDificultad() {
        int valor;
        do {
            System.out.println("---- Eligen la Dificultad ----");
            System.out.println("1.- Muy Facil.");
            System.out.println("2.- Facil.");
            System.out.println("3.- Medio.");
            System.out.println("4.- Dificil.");
            System.out.println("5.- Muy Dificil.");
            System.out.print("\nElige Opción: ");
            valor = sc_int.nextInt();

            if (valor < 1 || valor > 5) {
                System.out.println("** Error, elija una dificultad correcta.");
            }
        } while (valor < 1 || valor > 5);

        return valor;
    }
    
    public static void Avisos(int cual){
        switch(cual){
            case 1: //Sudoku creado correctamente
                System.out.println("*** Nuevo Sudoku creado correctamente. ***\n");
                break;
            case 2: //Sudoku cargado correctamente
                System.out.println("*** Sudoku cargado correctamnete. ***\n");
                break;
            case 3: //Sudoku borrado correctamente
                System.out.println("*** Sudoku eliminado correctamente. ***\n");
                break;
            case 4: //Hueco ocupado
                System.out.println("*** La posicion seleccionada ya está ocupada. ***\n");
                break;
            case 5: //Hueco rellenado correctamente
                System.out.println("*** Valor ingresado correctamente. ***\n");
                break;
            case 6: //Valor no permitido 
                System.out.println("*** Ingresa un valor valido. ***\n");
                break;
            case 7: //Lista de Sudokus vacia 
                System.out.println("*** No hay ningún sudoku que mostrar. ***\n");
                break;
            case 8: //Valor borrado 
                System.out.println("*** Valor borrado con exito. ***\n");
                break;
            case 9: //Mostrar Ayuda
                System.out.println("*** Los posibles valores en la posicion F=" + fila_aux + " C=" + col_aux + " son: " + Ayuda + "\n");
                break;
            case 10: //Penalizacion ayuda
                System.out.println("*** Los posibles valores en la posicion F=" + fila_aux + " C=" + col_aux + " son: " + Ayuda);
                System.out.println("*** Como penalización se le ha borrado un número aleatorio del Sudoku\n");
                break;
        }
    }

    public static void main(String[] args) {
        try {
            int opcPrin = -1, opcDif = -1, opcGest = -1, opcJug = -1, codActual = 0, seleccion, puerto;
            int fil = -1, col = -1, valor = -1;
            TipoDificultad Dificultad = TipoDificultad.VACIO;
            
            System.out.print("Introduce el num. de puerto: "); //Usar el 5500
            puerto = sc_int.nextInt();
            
            Registry registro = LocateRegistry.getRegistry(puerto);
            GestorJuegosIntf srv = (GestorJuegosIntf)registro.lookup("GestorDeJuegos");
            
            do {                
                System.out.println("\nCódigo: " + codActual + "\t" + srv.GetSudoku(codActual));
                opcPrin = MenuPrincipal();
                switch (opcPrin){
                    case 1: //Al Menu de Gestión
                        do {                            
                            System.out.println("\n\nCódigo: " + codActual + "\t" + srv.GetSudoku(codActual));
                            Avisos(aviso);
                            opcGest = MenuGestion();
                            switch (opcGest) {
                                case 1: //Crear Juego
                                    opcDif = MenuDificultad();
                                    switch(opcDif){
                                        case 1:
                                            Dificultad = TipoDificultad.MUY_FACIL;
                                            break;
                                        case 2:
                                            Dificultad = TipoDificultad.FACIL;
                                            break;
                                        case 3:
                                            Dificultad = TipoDificultad.MEDIA;
                                            break;
                                        case 4:
                                            Dificultad = TipoDificultad.DIFICIL;
                                            break;
                                        case 5:
                                            Dificultad = TipoDificultad.MUY_DIFICIL;
                                            break;
                                    }
                                    codActual = srv.NuevoJuego(Dificultad);
                                    ListaSudoku ls = new ListaSudoku(codActual, Dificultad);
                                    Sudokus.add(ls);
                                    aviso = 1;
                                    break;

                                case 2: //Borrar Juego (llama a ListarJuegos)
                                    if(!Sudokus.isEmpty()){
                                        seleccion = ListarJuegos(true); //true = Borrar
                                        srv.BorrarJuego(Sudokus.get(seleccion).Codigo);
                                        Sudokus.remove(seleccion);
                                        aviso = 3;
                                    }
                                    else
                                        aviso = 7;
                                    break;

                                case 3: //Seleccionar Juego (llama a ListarJuegos)
                                    if(!Sudokus.isEmpty()){
                                        seleccion = ListarJuegos(false); //false = seleccionar
                                        codActual = Sudokus.get(seleccion).Codigo;
                                        aviso = 2;
                                    }
                                    else
                                        aviso = 7;
                                    break;

                                case 0: //Volver al menú principal
                                    aviso = 0;
                                    break;

                                default:
                                    System.out.println("\n** Error, indica una opcion válida **");
                            }
                        } while (opcGest != 0);
                        break; //------------------> FIN CASE 1
                        
                    case 2: //Al Menu de Juegos
                        do {                            
                            System.out.println("\n\nCódigo: " + codActual + "\t" + srv.GetSudoku(codActual));
                            if (srv.NumeroHuecos(codActual) == 0) {
                                System.out.print("*** SUDOKU COMPLETO ");
                                if (srv.Correcto(codActual) == false) {
                                    System.out.print("PERO IN");
                                }
                                System.out.println("CORRECTO ***\n");
                            }
                            Avisos(aviso);
                            opcJug = MenuJuego();
                            switch (opcJug) {
                                case 1: //Poner Valor
                                    System.out.print("Indica la fila: ");
                                    fil = sc_int.nextInt();
                                    System.out.print("Indica la columna: ");
                                    col = sc_int.nextInt();
                                    System.out.print("Indica el valor: ");
                                    valor = sc_int.nextInt();

                                    if (fil <= 0 || fil > 9 || col <= 0 || col > 9 || valor <= 0 || valor > 9)
                                        aviso = 6;
                                    else{
                                        if(srv.ComprobarValor(codActual, fil-1, col-1, Integer.toString(valor).charAt(0)))
                                            aviso = 4;
                                        else {
                                            srv.PonerValor(codActual, fil - 1, col - 1, Integer.toString(valor).charAt(0));
                                            aviso = 5;
                                        }
                                    }
                                    break;

                                case 2: //Borrar Valor
                                    System.out.print("Indica la fila: ");
                                    fil = sc_int.nextInt();
                                    System.out.print("Indica la columna: ");
                                    col = sc_int.nextInt();

                                    if (fil <= 0 || fil > 9 || col <= 0 || col > 9) 
                                        aviso = 6;
                                    else{
                                        srv.PonerValor(codActual, fil - 1, col - 1, ' ');
                                        aviso = 8;
                                    }
                                    break;

                                case 3: //Ayuda
                                    System.out.print("Indica la fila: ");
                                    fila_aux = sc_int.nextInt();
                                    System.out.print("Indica la columna: ");
                                    col_aux = sc_int.nextInt();
                                    Ayuda = srv.Ayuda(codActual, fila_aux-1, col_aux-1);
                                    if(srv.NumeroHuecos(codActual) != 81)
                                        aviso = 10;
                                    else
                                        aviso = 9;
                                    break;

                                case 0: //Volver al menú principal
                                    aviso = 0;
                                    break;
                            }
                        } while (opcJug != 0);
                        break; //------------------> FIN CASE 2
                        
                    case 0: //Salir de la aplicación
                        break;
                        
                        default:
                            System.out.println("\n** Error, indica una opcion válida **");
                }
            } while (opcPrin != 0);
            
        } catch (RemoteException ex) {
            System.out.println("\n** ERROR, excepcion en el Cliente_Sudoku: " + ex.getMessage());
        } catch (NotBoundException ex) {
            Logger.getLogger(Cliente_Sudoku.class.getName()).log(Level.SEVERE, null, ex);
        }
    }//Fin del main

}

File: 3er CURSO/2-CUATRI/SD/PRACTICAS/Práctica 4 (C#)/GestorJuegos_MOD/ClienteSudoku/ClienteSudoku.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using System.Runtime.Remoting;
using System.Runtime.Remoting.Channels;
using System.Runtime.Remoting.Channels.Tcp;

namespace GestorJuegos
{
    class ClienteSudoku
    {
        static List<ListaSudoku> Sudokus = new List<ListaSudoku>();
        static int aviso = 0;
        static String Ayuda = "";
        static int fila_aux, col_aux;

        static int menuPrincipal(){
            int valor;
            Console.WriteLine("---- Servidor de Juegos ----");
            Console.WriteLine("1.- Gestionar Juegos.");
            Console.WriteLine("2.- Jugar.");
            Console.WriteLine("0.- Salir.");
            Console.Write("\nElige Opción: ");
            int.TryParse(Console.ReadLine(), out valor);

            return valor;
        }

        //======== MODIFICACIÓN ============
        static int menuGestion(){
            int valor;
            Console.WriteLine("---- Gestión de Juegos ----");
            Console.WriteLine("1.- Crear Juego.");
            Console.WriteLine("2.- Borrar Juego.");
            Console.WriteLine("3.- Seleccionar Juego.");
            Console.WriteLine("4.- Clonar Juego.");
            Console.WriteLine("0.- Salir.");
            Console.Write("\nElige Opción: ");
            int.TryParse(Console.ReadLine(), out valor);

            return valor;
        }

        static int menuJuego()
        {
            int valor;
            Console.WriteLine("---- Menú de Juegos ----");
            Console.WriteLine("1.- Poner Valor.");
            Console.WriteLine("2.- Borrar Valor.");
            Console.WriteLine("3.- Ayuda.");
            Console.WriteLine("0.- Salir.");
            Console.Write("\nElige Opción: ");
            int.TryParse(Console.ReadLine(), out valor);

            return valor;
        }

        //======== MODIFICACIÓN ============
        public static int ListarJuegos(int selección)
        {
            int valor;
            Console.WriteLine("\n\nLista de Juegos Almacenados\n");
            Console.WriteLine("Posición        Código          Dificultad");
            Console.WriteLine("------------------------------------------");
            int j = 1;
            for (int i = 0; i < Sudokus.Count; i++)
            {
                Console.WriteLine(j++ + "\t\t" + Sudokus[i].getCodigo() + "\t\t" + Sudokus[i].GetDificultad());
            }
            if (selección == 1) //Borrar
                Console.Write("\nIntroduce la posición del Juego a borrar: ");

            else if(selección == 2) //Clonar
                Console.Write("\nIntroduce la posición del Juego a clonar: ");

            else //Seleccionar
                Console.Write("\nIntroduce la posición del Juego a continuar: ");

            int.TryParse(Console.ReadLine(), out valor);

            return valor - 1;
        }

        public static int MenuDificultad()
        {
            int valor;
            do
            {
                Console.WriteLine("---- Eligen la Dificultad ----");
                Console.WriteLine("1.- Muy Facil.");
                Console.WriteLine("2.- Facil.");
                Console.WriteLine("3.- Medio.");
                Console.WriteLine("4.- Dificil.");
                Console.WriteLine("5.- Muy Dificil.");
                Console.Write("\nElige Opción: ");
                int.TryParse(Console.ReadLine(), out valor);

                if (valor < 1 || valor > 5)
                    Console.WriteLine("** Error, elija una dificultad correcta.");
                
            } while (valor < 1 || valor > 5);

            return valor;
        }

        //======== MODIFICACIÓN ============
        public static void Avisos(int cual)
        {
            switch (cual)
            {
                case 1: //Sudoku creado correctamente
                    Console.WriteLine("*** Nuevo Sudoku creado correctamente. ***\n");
                    break;
                case 2: //Sudoku cargado correctamente
                    Console.WriteLine("*** Sudoku cargado correctamnete. ***\n");
                    break;
                case 3: //Sudoku borrado correctamente
                    Console.WriteLine("*** Sudoku eliminado correctamente. ***\n");
                    break;
                case 4: //Hueco ocupado
                    Console.WriteLine("*** La posicion seleccionada ya está ocupada. ***\n");
                    break;
                case 5: //Hueco rellenado correctamente
                    Console.WriteLine("*** Valor ingresado correctamente. ***\n");
                    break;
                case 6: //Valor no permitido 
                    Console.WriteLine("*** Ingresa un valor valido. ***\n");
                    break;
                case 7: //Lista de Sudokus vacia 
                    Console.WriteLine("*** No hay ningún sudoku que mostrar. ***\n");
                    break;
                case 8: //Valor borrado 
                    Console.WriteLine("*** Valor borrado con exito. ***\n");
                    break;
                case 9: //Mostrar Ayuda
                    Console.WriteLine("*** Los posibles valores en la posicion F=" + fila_aux + " C=" + col_aux + " son: " + Ayuda + "\n");
                    break;
                case 10: //Opcion no valida
                    Console.WriteLine("*** Error, Indica una opción válida\n");
                    break;
                case 11: //Juego clonado
                    Console.WriteLine("*** Juego clonado correctamente ***\n");
                    break;
            }
        }

        static void Main(string[] args)
        {
            int opcPrin = -1, opcDif = -1, opcGest = -1, opcJug = -1, codActual = 0, seleccion, posSudoAct = -1;
            int fil = -1, col = -1, valor = -1;
            char OKborrar;
            TipoDificultad Dificultad = TipoDificultad.VACIO;

            //### Conexión con el servidor ###
            ChannelServices.RegisterChannel(new TcpChannel(), false);
            GestorJuegos gestor = (GestorJuegos)Activator.GetObject(typeof(GestorJuegos), "tcp://localhost:12345/GestorJuegos");

            do
            {
                Console.Clear();
                Console.WriteLine("\nCódigo: " + codActual + "\t" + gestor.GetSudoku(codActual));
                Avisos(aviso);
                opcPrin = menuPrincipal();

                switch (opcPrin)
                {
                    case 1: //Al menú de gestión
                        do
                        {
                            Console.Clear();
                            Console.WriteLine("\nCódigo: " + codActual + "\t" + gestor.GetSudoku(codActual));
                            Avisos(aviso);
                            opcGest = menuGestion();
                            switch (opcGest)
                            {
                                case 1: //Crear Juego
                                    Console.Clear();
                                    opcDif = MenuDificultad();
                                    switch (opcDif)
                                    {
                                        case 1:
                                            Dificultad = TipoDificultad.MUY_FACIL;
                                            break;
                                        case 2:
                                            Dificultad = TipoDificultad.FACIL;
                                            break;
                                        case 3:
                                            Dificultad = TipoDificultad.MEDIA;
                                            break;
                                        case 4:
                                            Dificultad = TipoDificultad.DIFICIL;
                                            break;
                                        case 5:
                                            Dificultad = TipoDificultad.MUY_DIFICIL;
                                            break;
                                    }
                                    codActual = gestor.NuevoJuego(Dificultad);
                                    ListaSudoku ls = new ListaSudoku(codActual, Dificultad);
                                    Sudokus.Add(ls);
                                    posSudoAct = Sudokus.Count - 1;
                                    aviso = 1;
                                    break;

                                case 2: //Borrar Juego
                                    if (codActual != 0){
                                        Console.Write("¿Desea borrar el Sudoku actual? (s/n): ");
                                        char.TryParse(Console.ReadLine(), out OKborrar);
                                        if (OKborrar == 's' || OKborrar == 'S')
                                        {
                                            gestor.BorrarJuego(codActual);
                                            codActual = 0;
                                            Sudokus.RemoveAt(posSudoAct);
                                            aviso = 3;
                                        }
                                        else{
                                            Console.Clear();
                                            seleccion = ListarJuegos(1); //1 = Borrar
                                            if (seleccion >= 0 && seleccion < Sudokus.Count)
                                            {
                                                gestor.BorrarJuego(Sudokus[seleccion].getCodigo());
                                                if (seleccion == posSudoAct)
                                                    codActual = 0;
                                                Sudokus.RemoveAt(seleccion);
                                                aviso = 3;
                                            }
                                            else
                                                aviso = 0;
                                        }
                                    }
                                    else {
                                        if (Sudokus.Count > 0)
                                        {
                                            Console.Clear();
                                            seleccion = ListarJuegos(1); //1 = Borrar
                                            if (seleccion >= 0 && seleccion < Sudokus.Count)
                                            {
                                                gestor.BorrarJuego(Sudokus[seleccion].getCodigo());
                                                if (seleccion == posSudoAct)
                                                    codActual = 0;
                                                Sudokus.RemoveAt(seleccion);
                                                aviso = 3;
                                            }
                                            else
                                                aviso = 0;
                                        }
                                        else
                                            aviso = 7;
                                    }
                                    break;

                                case 3: //Seleccionar Juego
                                    if (Sudokus.Count != 0){
                                        Console.Clear();
                                        seleccion = ListarJuegos(0); //0 = Seleccionar
                                        if (seleccion >= 0 && seleccion < Sudokus.Count)
                                        {
                                            codActual = Sudokus[seleccion].getCodigo();
                                            posSudoAct = seleccion;
                                            aviso = 2;
                                        }
                                        else
                                            aviso = 0;    
                                    }
                                    else
                                        aviso = 7;
                                    break;

                                //======== MODIFICACIÓN ============
                                case 4: //Clonar Juego
                                    if (Sudokus.Count != 0){
                                        Console.Clear();
                                        seleccion = ListarJuegos(2); //0 = Clonar
                                        if (seleccion >= 0 && seleccion < Sudokus.Count)
                                        {
                                            codActual = gestor.ClonarJuego(Sudokus[seleccion].getCodigo());
                                            ListaSudoku aux = new ListaSudoku(codActual, Sudokus[seleccion].GetDificultad());
                                            Sudokus.Add(aux);
                                            posSudoAct = Sudokus.Count-1;
                                            aviso = 11;
                                        }
                                        else
                                            aviso = 0;
                                    }
                                    else
                                        aviso = 7;
                                    break;

                                case 0: //Salir
                                    aviso = 0;
                                    break;

                                default:
                                    aviso = 10;
                                    break;
                            }
                        } while (opcGest != 0);
                        break;

                //------------------------------------------------------------------------------
                    case 2: //Al menú de juego
                        do
                        {
                            Console.Clear();
                            Console.WriteLine("\nCódigo: " + codActual + "\t" + gestor.GetSudoku(codActual));
                            if (gestor.NumeroHuecos(codActual) == 0){
                                Console.Write("*** SUDOKU COMPLETO ");
                                if (gestor.Correcto(codActual) == false)
                                    Console.Write("PERO IN");
                                
                                Console.WriteLine("CORRECTO ***\n");
                            }
                            Avisos(aviso);
                            
                            opcJug = menuJuego();
                            switch (opcJug)
                            {
                                case 1: //Poner Valor
                                    Console.Write("Indica la fila: ");
                                    int.TryParse(Console.ReadLine(), out fil);
                                    Console.Write("Indica la columna: ");
                                    int.TryParse(Console.ReadLine(), out col);
                                    Console.Write("Indica el valor: ");
                                    int.TryParse(Console.ReadLine(), out valor);

                                    if (fil <= 0 || fil > 9 || col <= 0 || col > 9 || valor <= 0 || valor > 9)
                                        aviso = 6;
                                    else{
                                        if (gestor.ComprobarValor(codActual, fil - 1, col - 1, valor.ToString()[0]))
                                            aviso = 4;
                                        
                                        else{
                                            gestor.PonerValor(codActual, fil - 1, col - 1, valor.ToString()[0]);
                                            aviso = 5;
                                        }
                                    }
                                    break;

                                case 2: //Borrar Valor
                                    Console.Write("Indica la fila: ");
                                    int.TryParse(Console.ReadLine(), out fil);
                                    Console.Write("Indica la columna: ");
                                    int.TryParse(Console.ReadLine(), out col);

                                    if (fil <= 0 || fil > 9 || col <= 0 || col > 9)
                                        aviso = 6;
                                    else{
                                        gestor.PonerValor(codActual, fil - 1, col - 1, ' ');
                                        aviso = 8;
                                    }
                                    break;

                                case 3: //Ayuda
                                    Console.Write("Indica la fila: ");
                                    int.TryParse(Console.ReadLine(), out fila_aux);
                                    Console.Write("Indica la columna: ");
                                    int.TryParse(Console.ReadLine(), out col_aux);
                                    Ayuda = gestor.Ayuda(codActual, fila_aux - 1, col_aux - 1);
                                    aviso = 9;
                                    break;

                                case 0: //Volver al menú principal
                                    aviso = 0;
                                    break;

                                default:
                                    aviso = 10;
                                    break;
                            }
                        } while (opcJug != 0);
                        break;

                //------------------------------------------------------------------------------
                    case 0: //Salir
                        aviso = 0;
                        break;

                    default:
                        aviso = 10;
                        break;
                }
            } while (opcPrin != 0);
        }
    }
}
